src/CVRF/cvrf_priv.c

```diff
@@ -0,0 +1,2292 @@
+	branch->branch_name = (char *)xmlTextReaderGetAttribute(reader, TAG_NAME);
+	branch->branch_type = cvrf_item_parse_type_attribute(reader, CVRF_BRANCH);
+	xmlTextReaderNextElement(reader);
+
+	if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_PRODUCT_NAME)) {
+		branch->full_name = cvrf_product_name_parse(reader);
+		xmlTextReaderNextNode(reader);
+		xmlTextReaderNextNode(reader);
+	}
+	else {
+		while(xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_BRANCH) == 0) {
+			if (xmlTextReaderNodeType(reader) != XML_READER_TYPE_ELEMENT) {
+				xmlTextReaderNextNode(reader);
+				continue;
+			}
+			if (!oscap_list_add(branch->subbranches, cvrf_branch_parse(reader))) {
+				cvrf_set_parsing_error("Branch");
+			}
+			xmlTextReaderNextNode(reader);
+		}
+	}
+	xmlTextReaderNextNode(reader);
+	return branch;
+}
+
+struct cvrf_relationship *cvrf_relationship_parse(xmlTextReaderPtr reader) {
+	__attribute__nonnull__(reader);
+
+	struct cvrf_relationship *relation = cvrf_relationship_new();
+	relation->product_reference = (char *)xmlTextReaderGetAttribute(reader, ATTR_PRODUCT_REFERENCE);
+	relation->relation_type = cvrf_item_parse_type_attribute(reader, CVRF_RELATIONSHIP);
+	relation->relates_to_ref = (char *)xmlTextReaderGetAttribute(reader, ATTR_RELATES_TO_REF);
+	xmlTextReaderNextElement(reader);
+
+	if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_PRODUCT_NAME)) {
+		relation->full_name = cvrf_product_name_parse(reader);
+	}
+	xmlTextReaderNextNode(reader);
+
+	return relation;
+}
+
+struct cvrf_group *cvrf_group_parse(xmlTextReaderPtr reader) {
+	__attribute__nonnull__(reader);
+
+	struct cvrf_group *group = cvrf_group_new();
+	group->group_id = (char *)xmlTextReaderGetAttribute(reader, TAG_GROUP_ID);
+	xmlTextReaderNextElement(reader);
+
+	while (xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_GROUP) != 0) {
+		if (xmlTextReaderNodeType(reader) != XML_READER_TYPE_ELEMENT) {
+			xmlTextReaderNextNode(reader);
+			continue;
+		}
+		if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_DESCRIPTION)) {
+			group->description = oscap_element_string_copy(reader);
+		} else if (!xmlStrcmp(xmlTextReaderConstLocalName(reader), TAG_PRODUCT_ID)) {
+			oscap_stringlist_add_string(group->product_ids, oscap_element_string_copy(reader));
+		}
+		xmlTextReaderNextNode(reader);
+	}
+	xmlTextReaderNextNode(reader);
+	return group;
+}
+
+struct cvrf_product_name *cvrf_product_name_parse(xmlTextReaderPtr reader) {
+	struct cvrf_product_name *full_name = cvrf_product_name_new();
+	full_name->product_id = (char *)xmlTextReaderGetAttribute(reader, TAG_PRODUCT_ID);
+	full_name->cpe = oscap_element_string_copy(reader);
+	xmlTextReaderNextNode(reader);
+	return full_name;
+}
+
+struct cvrf_vulnerability *cvrf_vulnerability_parse(xmlTextReaderPtr reader) {
+	__attribute__nonnull__(reader);
+
+	struct cvrf_vulnerability *vuln = cvrf_vulnerability_new();
+	vuln->ordinal = (int)oscap_strtol((char *)xmlTextReaderGetAttribute(reader, ATTR_ORDINAL), NULL, 10);
```

Comment:
IMO just use the C standard builtin, make sure it's actually in the standard and if it is just use it...